### ISSUE TITLE
Expected way to include functionality to objects

### ISSUE THREAD

**Issue by hannestschofenig** [eclipse/wakaama#638]
Imaging the following use case: I would like to use Wakaama to perform a firmware update and thereby use the firmware update object. The Wakaama client contains the object_firmware.c but there is no code that actually does the firmware update.

What is the expected way to add this necessary code?

Would I write it directly into object_firmware.c (which would be easy)? If so, re-usability becomes a challenge.

**Comment by LukasKarel** [comment]
In my opinion it depends.
You cant describe a firmware update process which is usable for all targets (Linux-based client, bare metal microcontroller without OS, ...). Therefore you must implement the firmware update process in a own object_firmware.c specific for your target.

**Comment by hannestschofenig** [comment]
So, your proposal is to create a new object_firmware.c file that implements the necessary functionality and to leave the existing file as is. 

That's a possible approach but will lead to a situation where nobody contributes platform-specific code to Wakaama, which makes Wakaama usable only for educational purposes rather than something that can be used for actual deployments. 

There are projects that follow a different approach. For example, the Mcuboot project implements a bootloader and it can be run on existing hardware while at the same time offering an abstract level that remains unchanged from the hardware. The consequence is that you can actually run it on hardware and if you need to port it to an unsupported hardware platform you can use the existing code as a reference.

**Comment by LukasKarel** [comment]
I agree with you. But at the current state of the project there is no useable abstraction (except networking). And as the development of wakaama is inactive again I wouldnt expect that this will change in the near future.

**Comment by LukasWoodtli** [comment]
Closing because the Wakaama project will be terminated: See issue #883 and [iot.wakaama 2026.05 Termination Review](https://gitlab.eclipse.org/eclipsefdn/emo-team/emo/-/issues/1191)
